### DISCUSSION TITLE
Unable to run anything from the Getting Started guidance on apple silicon

### DISCUSSION THREAD

**Discussion by JamesHenry** [napi-rs/napi-rs#727] [Q&A]
Hi there,

I opened this as a discussion rather than an issue because I can't help feeling like I must be misunderstanding on my end (I am freely willing to admit that architecture + native compiler concerns are outside my wheelhouse).

Thank you so much in advance for any guidance you give 🙏 

(For reference, I am able to build and run Rust programs via Cargo on this same machine without any problems)

**Env details:**
- I am on a Mac Mini, so using the M1 Apple Silicon hardware
- I usually use https://volta.sh to manage my node installation, but even uninstalled that entirely and tried a completely fresh node installation from the official website to remove all possible variables there.
- I tried with both node 14 and node 16, same results

**What I am experiencing**
- I tried both the CLI new option, and using the Github based template repo - both yield the same results

- When I run `yarn build/npm run build` it fails saying that the module `iconv-lite` is missing. Again, this feels like it would be so obviously broken that I don't understand why everyone isn't running into this so must be something on my end? (And again this happens with both the CLI new and your template)

- If I manually install `iconv-lite` to move past this the build succeeds and produces e.g. `package-template.darwin-arm64.node`
- If I then run `node index.js`, or use your `simple-test.js` which requires in `index.js`, I get `Can not load bindings` with no more information:

```sh
❯ node index.js                                                                                                                                                                                 х 1
/Users/james/Repos/private/jameshenry/rusty-nx-project-graph/node_modules/@node-rs/helper/lib/loader.js:50
    throw new Error(errorMsg);
    ^

Error: Can not load bindings

    at loadBinding (/Users/james/Repos/private/jameshenry/rusty-nx-project-graph/node_modules/@node-rs/helper/lib/loader.js:50:11)
    at Object.<anonymous> (/Users/james/Repos/private/jameshenry/rusty-nx-project-graph/index.js:11:18)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
``` 

- If I create a JS file which brings in the `.node` file directly I get:

```sh
❯ node test.js                                                                                                                                                                                  х 1
internal/modules/cjs/loader.js:1131
  return process.dlopen(module, path.toNamespacedPath(filename));
                 ^

Error: dlopen(/Users/james/Repos/private/jameshenry/rusty-nx-project-graph/rusty-nx-project-graph.darwin-arm64.node, 1): no suitable image found.  Did find:
        /Users/james/Repos/private/jameshenry/rusty-nx-project-graph/rusty-nx-project-graph.darwin-arm64.node: mach-o, but wrong architecture
        /Users/james/Repos/private/jameshenry/rusty-nx-project-graph/rusty-nx-project-graph.darwin-arm64.node: mach-o, but wrong architecture
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1131:18)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/Users/james/Repos/private/jameshenry/rusty-nx-project-graph/test.js:1:11)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
```

**Comment by Brooooooklyn** [comment]
Checkout your Node.js arch, seems like your Node.js is `x86_64` binary and the addon is for `arm64`.

Just log `process.arch` for Node.js arch

**Reply by JamesHenry** [reply]
Yes you are right, it logs `x64`. I believe node hasn't supported native binaries for apple silicon (until recently, maybe?)

So how would I build the correct binary locally in your template project when running `npm run build`?

Any thoughts on the missing `iconv-lite`?

**Reply by JamesHenry** [reply]
(Thank you for the very prompt response btw)

**Reply by JamesHenry** [reply]
(P.S. If I use the latest version of node (16.8), I do get the log arch64 for the process.arch and simple-test.js will then execute on the built binary so that's great. Would still love your thoughts on building for x64 node versions and `iconv-lite` being missing)

**Reply by Brooooooklyn** [reply]
As far as I remember Node.js started supporting macOS-arm64 from some version of v15.
I guess the `iconv-lite` problem is because volta didn't deliver `x64 npm client` right to your machine, which may dependent on `iconv-lite`.

> So how would I build the correct binary locally in your template project when running npm run build

As I know, macOS-arm64 can't cross compile `x64` binary, while macOS-x64 can cross compile `arm64` binary.

**Reply by Brooooooklyn** [reply]
Oh sorry, the `iconv-lite` problem is because the wrong bundle config in `@napi-rs/cli`, it should be fixed in [1.3.1](https://github.com/napi-rs/napi-rs/releases/tag/%40napi-rs%2Fcli%401.3.1)

**Reply by JamesHenry** [reply]
> Oh sorry, the iconv-lite problem is because the wrong bundle config in @napi-rs/cli, it should be fixed in 1.3.1

Ah ok, great! Glad I wasn't going crazy with that one :)

I can confirm reinstalling the latest version of the CLI and creating a new project does not have any of the same issues.

> As I know, macOS-arm64 can't cross compile x64 binary, while macOS-x64 can cross compile arm64 binary.

Ok interesting, so my approach should be - build for myself on the m1 for development purposes (always using a recent version of node so that I have a native m1 build of node itself), and let the generated github actions handle building for x64 macs?

**Reply by Brooooooklyn** [reply]
> so my approach should be - build for myself on the m1 for development purposes (always using a recent version of node so that I have a native m1 build of node itself), and let the generated github actions handle building for x64 macs?

Yes, let CI build all the other platforms binary for you
